Apple Help Book generation

al45tair opened the pull request:
This pull request adds support for generating Apple Help Books to Sphinx.

al45tair committed:
Added initial Apple help support.

al45tair committed:
More changes for Apple Help support.

al45tair committed:
Added code signing support.  Improved error handling.  Fixed handle_finish to work as it's supposed to.

al45tair committed:
Clean up the mechanism for disabling Javascript search.

al45tair committed:
Disable Javascript search in Apple Help.

al45tair committed:
Altered the way we use applehelp_locale slightly.  Also altered its default setting.

al45tair committed:
Added documentation for Apple Help builder.

al45tair committed:
Fixed a typo.

al45tair committed:
Encode the project name using IDNA if necessary.

al45tair committed:
Added myself to AUTHORS.

birkenfeld commented on doc/config.rst:
```diff
@@ -899,6 +899,139 @@ that use Sphinx's HTMLWriter class.
+
+Options for Apple Help output
+-----------------------------
+
```
The section should get a versionadded for 1.3.

birkenfeld commented on doc/builders.rst:
```diff
@@ -76,6 +76,22 @@ The builder's "name" must be given to the **-b** command-line option of
+   This builder produces an Apple Help Book based on the same output as the
+   standalone HTML builder.
+
+   This builder requires Mac OS X 10.6 or higher to function, as it depends on
```
E.g. for the Windows HTML help builder, we have omitted calling the "help compiler" from Sphinx and added that to the instructions of things to do. That way, people can use it somewhere else or use a non-standard (i.e. Wine) installation. Would it make sense to do a similar thing here?

At least for testing, the builder should be able to run all but the "run external programs" steps on all platforms.

al45tair replied:
The reason I opted to call them directly is that building an Apple Help bundle is mostly just a case of assembling the collection of files in the right places, then running the indexer to add a help index (and optionally, code-signing). This is in contrast to e.g. the situation with Windows HTML Help, where the files need to be assembled as a help project and then processed by the help compiler to make a special archive file. I think it makes more sense for Sphinx to be able to complete all the necessary steps, since there isn't a special tool that does so.

That said, you’re right that it would be good to have some support for either using a different `hiutil` in a different path, or not generating the index at all and leaving it for later.

birkenfeld commented on sphinx/quickstart.py:
```diff
@@ -266,6 +268,58 @@
 # Output file base name for HTML help builder.
 htmlhelp_basename = '%(project_fn)sdoc'
 
+# -- Options for Apple Help output ----------------------------------------
```
We haven't added potential options for other "minor" builders to the quickstart config, such as devhelp or qthelp, to keep the length from exploding. Since your config values have reasonable defaults, I would prefer that here too.

al45tair replied:
OK. I'll remove those. I'm also going to remove the default for `applehelp_bundle_id` because I think on reflection, if it isn't present in the quick start output, people won't set it and we’ll end up with bad bundle IDs in shipping help files.

birkenfeld commented:
Thanks for the PR! The new builder isn't tested at all - the minimum amount I'd like to have is to build the test project with it.

See tests/test_build.py:test_build_all.

al45tair committed:
Added some additional configuration variables for Apple Help.
Removed Apple Help from quickstart conf.py.
Added support for .lproj directories with pre-localised files for Apple Help.

al45tair committed:
Added tests for Apple Help generation.

al45tair committed:
Don't use codecs for writing plists; breaks on Python 3.

al45tair committed:
Use binary mode for writing plist file, as per Python 3 docs.

al45tair committed:
Use binary mode for reading plist file, as per Python 3 docs.

al45tair commented:
OK, I’ve added configuration variables to let users pick an alternate `hiutil` or `codesign` executable, or to disable external tool use altogether. I also added some tests, and fixed a couple of Python 3 issues I spotted. Finally, I’ve added support for `<lang>.lproj` folders in the source directory, which are ignored by the other builders but when you use the apple help builder will be merged with the output (this is to handle things like `InfoPlist.strings`, which isn’t dealt with via the existing Sphinx localisation system).

birkenfeld commented:
From a cursory review, this looks good now! I'll merge it for 1.3. Thanks for the contribution!
